Fix #5893: Invalidate widget only if it's visible

w-flo opened the pull request:
When reviewing, please take into account that I have looked at OpenRCT2 code for the first time a few days ago, so maybe this pull request is terrible. This fixes #5893 for me and I have verified using the software and OpenGL renderers that action labels in guest windows update correctly when guests leave a ride, then head to the same ride again and either re-enter the queue or fail to get in line because the queue is full. However, I'm having a hard time to verify that this doesn't break anything, because the action label seems to update without any invalidate calls anyway, maybe because of the viewport in the guest window that updates all the time and is so close to the label?

Note:
- This removes calls to "window_event_invalidate_call(w);". As far as I can tell, that function executes an event callback like [window_guest_overview_invalidate()](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2-ui/windows/Guest.cpp#L1031), and that just fixes the guest window layout. The layout shouldn't change when a guest decides to head to some ride (or stops heading to a ride), so I thought that call could be dropped. If it can't, it could probably be added inside the new if body in window_guest_overview_update().
- This changes the peep->window_invalidate_flags value for most guests by setting one bit to 1 that was probably 0 before (unless the guest window for that guest is open). As far as I can tell, that value is stored in save files. It shouldn't cause issues in OpenRCT2, but I can't test / tell if it causes any issues in vanilla RCT2 when loading a save game that has this flag set for guests. The field name in OpenRCT2 sounds so innocent, but that's just OpenRCT2 I guess?

Also, the flag name is a bit weird, considering that it is only for "heading_to_ride has changed" right  now. When looking at it as a way for libopenrtc2 to signal to the -ui project that it should invalidate the action label it makes more sense (but the action label could change for other reasons and libopenrtc2 calls a full window_invalidate() on its own in those cases). It should be possible to use the flag in other places where libopenrtc2 currently calls window_invalidate() after changing the guest state, but I wanted to change only those cases where the existing code base calls widget_invalidate(), because that's where my game crashes because of the failed assertion.

Commit message:
Remove the widget_invalidate() call after changing a guest's
guest_heading_to_ride_id, because that call fails a debug assertion if
the guest window is open and the currently active tab is not the
"overview" tab. In Release builds (if assertion is disabled),
widget_invalidate() might access the widgets array out of bounds.

Instead, introduce a new flag PEEP_INVALIDATE_PEEP_ACTION for
window_invalidate_flags in the peep struct and set that flag. The guest
window update function then makes sure to invalidate the label if the
flag is set.

The flag could be used in other places to reduce libopenrct2 dependency
on window_*() calls (see #6808), but this commit only cares about cases
where the assertion would fail.

AaronVanGeffen commented on src/openrct2/peep/Peep.h:
```diff
@@ -511,6 +511,7 @@ enum PeepInvalidate
     PEEP_INVALIDATE_PEEP_2 = 1 << 2,
     PEEP_INVALIDATE_PEEP_INVENTORY = 1 << 3,
     PEEP_INVALIDATE_STAFF_STATS = 1 << 4,
+    PEEP_INVALIDATE_PEEP_ACTION = 1 << 5     // Currently set only when guest_heading_to_ride_id is changed
```
```suggestion
    PEEP_INVALIDATE_PEEP_ACTION = 1 << 5, // Currently set only when guest_heading_to_ride_id is changed
```

w-flo replied:
Whoops, turns out I should actually add that pre-commit hook instead of pasting the command and accidentally dropping the *.h part. This & contributors.md is in the second commit (feel free to squash both commits into one).

AaronVanGeffen replied:
You've accidentally left two spaces between the comma and slash, though. ;) As a result, the clang-format CI job still fails.

w-flo replied:
Okay, this is now becoming embarrassing. The second space was a pre-commit hook test to make sure it works. And it worked! So I removed that second space again and committed, but forgot to "git add" the fixed file before committing. So I blame the pre-commit hook. :-)

AaronVanGeffen commented:
Nice work! Please feel free to add yourself to `contributors.md` as well.

duncanspumpkin commented on src/openrct2-ui/windows/Guest.cpp:
```diff
@@ -1090,6 +1089,8 @@ void window_guest_overview_invalidate(rct_window* w)
  */
 void window_guest_overview_update(rct_window* w)
 {
+    Peep* peep = GET_PEEP(w->number);
```
Move this closer to when it is used

duncanspumpkin commented on src/openrct2-ui/windows/Guest.cpp:
```diff
@@ -1098,6 +1099,12 @@ void window_guest_overview_update(rct_window* w)
     widget_invalidate(w, WIDX_TAB_1);
     widget_invalidate(w, WIDX_TAB_2);
 
+    if (peep->window_invalidate_flags & PEEP_INVALIDATE_PEEP_ACTION)
```
Do a nullptr check on peep.

duncanspumpkin commented:
Will need testing in multiplayer i think.

duncanspumpkin commented:
Oh also need to add a change log entry

w-flo commented:
Thanks for your comments!

> Will need testing in multiplayer i think.

That sounds important. I have never used multiplayer and I only have a vague idea how this would be tested. Like, is it enough to connect to a server and see if anything bad happens? Do I need to run my own server that has these changes, and then connect using a client with/without these changes? Should I simulate high network latency? Obviously I could try to do all these things, but I'm not sure if that is enough, I don't know how multiplayer is implemented in OpenRCT2 and which things this commit might break.

I will add a commit addressing your code comments and a changelog entry.

duncanspumpkin commented:
You just need to run two instances and make a local server and connect to that. Have one that has a peep window open so that it clears the invalidate flag. Just leave it running for a while and see if it desyncs. I'm pretty sure we already have code that prevents the invalidate flags from breaking multiplayer but i don't have the time right now to check.

w-flo commented:
It seems to work fine with 8 guest windows open in the server instance, and 0 windows in the connecting instance. Some of the 8 guests have already shown the "action label" message for heading_to_ride_id, so both instances should have different sets of peep window invalidate flags.

For fun, I decided to pick up peeps (both staff and guests) and sometimes this would result in a desync, but I assume that's unrelated to this PR and is simply #8731.

I also found this in the code for sprite_checksum(), which sounds promising:
```
                if (copy.generic.sprite_identifier == SPRITE_IDENTIFIER_PEEP)
                {
                    // We set this to 0 because as soon the client selects a guest the window will remove the
                    // invalidation flags causing the sprite checksum to be different than on server, the flag does not affect
                    // game state.
                    copy.peep.window_invalidate_flags = 0;
                }
```

duncanspumpkin commented:
Yup that was the bit of code i was thinking of. Think this is good to go then
